diff --git a/README.md b/README.md
index c69fdff..f1433fe 100644
--- a/README.md
+++ b/README.md
@@ -7,6 +7,7 @@ Format your Fennel!
     $ make install PREFIX=$HOME # or sudo make install in /usr/local
     $ ./fnlfmt mycode.fnl # prints formatted code to standard out
     $ ./fnlfmt --fix mycode.fnl # replaces the file with formatted code
+    $ ./fnlfmt --check mycode.fnl # checks if the file is formatted
     $ curl localhost:8080/my-file.fnl | ./fnlfmt - # pipe to stdin
 
 You can skip reformatting of top-level forms by placing a comment
diff --git a/cli.fnl b/cli.fnl
index 55d6c6d..520ec0b 100644
--- a/cli.fnl
+++ b/cli.fnl
@@ -5,10 +5,11 @@
 (local {: format-file : version} (require :fnlfmt))
 
 (fn help []
-  (print "Usage: fnlfmt [--no-comments] [--fix] FILENAME...")
-  (print "With the --fix argument, multiple files can be specified")
-  (print "and changes are made in-place; otherwise prints the")
-  (print "formatted file to stdout."))
+  (print "Usage: fnlfmt [--no-comments] [--fix] [--check] FILENAME...")
+  (print "With the --fix or --check argument, multiple files can be specified.")
+  (print "With the --fix argument, formats all the files in-place;")
+  (print "with the --check argument, only checks if all the files are formatted;")
+  (print "otherwise prints the formatted file to stdout."))
 
 (local options [])
 
@@ -17,6 +18,23 @@
     (set options.no-comments true)
     (table.remove arg i)))
 
+(fn check-file [filename]
+  (let [old (with-open [in (assert (io.open filename))] (in:read :*a))
+        new (format-file filename options)]
+    (or (= old new)
+        (do
+          (io.stderr:write (string.format "Not formatted: %s\n" filename))
+          false))))
+
+(fn check-files [filenames]
+  (let [ok? (accumulate [ok? true _ filename (pairs filenames)]
+              (and ok? (check-file filename)))]
+    (if ok?
+        (do
+          (io.stderr:write "All files are formatted.\n")
+          (os.exit true))
+        (os.exit false))))
+
 (fn fix [filename]
   (let [new (format-file filename options)
         f (assert (io.open filename :w))]
@@ -26,6 +44,7 @@
 (match arg
   [:--version] (print (.. "fnlfmt version " version))
   [:--fix & filenames] (each [_ filename (pairs filenames)] (fix filename))
+  [:--check & filenames] (check-files filenames)
   (where (or [:--help] ["-?"] [:-h])) (help)
   [filename nil] (io.stdout:write (format-file filename options))
   _ (help))
